### PR TITLE
Support new option "github-user"

### PR DESCRIPTION
**Pull request by yyoshiki41**
## Description

Add new option `github-user`.

## Motivation and Context

That allowed the specified github users that is not on team or org.
e.g.) Username list (comma-separated)
"github-user:octocat,yyoshiki41"

fixed below issue.
https://github.com/pusher/oauth2_proxy/issues/342

## How Has This Been Tested?

My develop environment works this option.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

---
This change is [Reviewable](https://reviewable.io/reviews/pusher/oauth2_proxy/421)

### CODE REVIEW

**Comment by tg44** [comment]
I'm not a pro, nor a maintainer, but this seems to be ok.

File: main.go [hunk 1]
```diff
@@ -64,6 +64,7 @@ func main() {
 	flagSet.String("bitbucket-repository", "", "restrict logins to user with access to this repository")
 	flagSet.String("github-org", "", "restrict logins to members of this organisation")
 	flagSet.String("github-team", "", "restrict logins to members of this team")
+	flagSet.String("github-user", "", "restrict logins to these users")
```
Comment (JoelSpeed) [inline main.go]: We have a number of options that allow the flag to be specified multiple times, eg `whitelist-domain`, would it be possible to follow the same pattern?

Comment (yyoshiki41) [reply main.go]: fixed by using `flagSet.StringSlice`.

**Comment by yyoshiki41** [comment]
@JoelSpeed 

Sorry for late reply.
I fixed some issues and pushed.
- [ ] rebase master
- [ ] fix tests
- [ ] add some tests
- [ ] update docs

Thanks!

File: options.go [hunk 2]
```diff
@@ -56,6 +56,7 @@ type Options struct {
 	GitHubTeam               string   `flag:"github-team" cfg:"github_team" env:"OAUTH2_PROXY_GITHUB_TEAM"`
 	GitHubRepo               string   `flag:"github-repo" cfg:"github_repo" env:"OAUTH2_PROXY_GITHUB_REPO"`
 	GitHubToken              string   `flag:"github-token" cfg:"github_token" env:"OAUTH2_PROXY_GITHUB_TOKEN"`
+	GitHubUsers              []string `flag:"github-user" cfg:"github_user" env:"OAUTH2_PROXY_GITHUB_USER"`
```
Comment (JoelSpeed) [inline options.go:59]: To be consistent with other list options, will also need to update the docs
```suggestion
	GitHubUsers              []string `flag:"github-user" cfg:"github_users" env:"OAUTH2_PROXY_GITHUB_USERS"`
```

File: providers/github.go [hunk 3]
```diff
@@ -80,6 +83,19 @@ func (p *GitHubProvider) SetRepo(repo, token string) {
...
+func (p *GitHubProvider) SetUsers(users []string) {
+	for _, user := range users {
+		us := strings.Split(user, ",")
+		for _, u := range us {
+			// Validate empty string (e.g. --github-user="octcat,,")
+			if len(u) > 0 {
+				p.Users = append(p.Users, u)
+			}
+		}
+	}
```
Comment (JoelSpeed) [inline providers/github.go:96]: This will be covered by viper, the configuration loader

File: providers/github.go [hunk 4]
```diff
@@ -317,6 +333,48 @@ func (p *GitHubProvider) hasRepo(ctx context.Context, accessToken string) (bool,
...
+		return false, fmt.Errorf("got %d from %q %s",
+			resp.StatusCode, stripToken(endpoint.String()), body)
+	}
+
+	if err := json.Unmarshal(body, &user); err != nil {
+		return false, err
+	}
+
+	if p.isVerifiedUser(user.Login) {
+		log.Printf("Found Github User: %q", user.Login)
```
Comment (JoelSpeed) [inline providers/github.go:371]: Will this log line become noisy? Same for the one below?

File: providers/github.go [hunk 5]
```diff
@@ -356,21 +414,38 @@ func (p *GitHubProvider) GetEmailAddress(ctx context.Context, s *sessions.Sessio
...
+			} else {
+				if ok, err := p.hasOrg(ctx, s.AccessToken); err != nil || !ok {
+					return "", err
+				}
 			}
-		} else {
-			if ok, err := p.hasOrg(ctx, s.AccessToken); err != nil || !ok {
-				return "", err
+		} else if p.Repo != "" {
+			if p.Token == "" { // If we have a token we'll do the collaborator check in GetUserName
```
Comment (JoelSpeed) [inline providers/github.go:443]: Why split these two up?

File: options.go (hunk 2 above)
Comment (yyoshiki41) [reply options.go]: https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/8952cfb01c99661f478465790477db58e29720b1
and updated docs: https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/be7f969902b11d75b0af877a153815d4fc912fdf.

**Comment by yyoshiki41** [comment]
@JoelSpeed 
Thanks for your review!
I fixed all your review points and pushed.

File: options.go [hunk 6]
```diff
@@ -56,6 +56,7 @@ type Options struct {
 	GitHubTeam               string   `flag:"github-team" cfg:"github_team" env:"OAUTH2_PROXY_GITHUB_TEAM"`
 	GitHubRepo               string   `flag:"github-repo" cfg:"github_repo" env:"OAUTH2_PROXY_GITHUB_REPO"`
 	GitHubToken              string   `flag:"github-token" cfg:"github_token" env:"OAUTH2_PROXY_GITHUB_TOKEN"`
+	GitHubUsers              []string `flag:"github-users" cfg:"github_users" env:"OAUTH2_PROXY_GITHUB_USERS"`
```
Comment (JoelSpeed) [inline options.go:59]: Sorry, I wasn't very clear in my previous comment. The flag should be singular, while the other two should be pluralised (yes this is very confusing, but it's what we have for now).
```suggestion
	GitHubUsers              []string `flag:"github-user" cfg:"github_users" env:"OAUTH2_PROXY_GITHUB_USERS"`
```

File: providers/github.go [hunk 7]
```diff
@@ -317,6 +325,47 @@ func (p *GitHubProvider) hasRepo(ctx context.Context, accessToken string) (bool,
...
+	}
+
+	if err := json.Unmarshal(body, &user); err != nil {
+		return false, err
+	}
+
+	if p.isVerifiedUser(user.Login) {
+		return true, nil
+	}
+	log.Printf("Missing Github User: %q in %v", user.Login, p.Users)
```
Comment (JoelSpeed) [inline providers/github.go:365]: I think we should drop this log too, could get noisy

File: options.go (hunk 6 above)
Comment (yyoshiki41) [reply options.go]: Thanks for your explanation, I understood!
https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/5dc11924b63c590f3b61f79ff0bac3a15107a138

File: providers/github.go (hunk 7 above)
Comment (yyoshiki41) [reply providers/github.go]: Remove logging https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/19f9b34ce854faf0e296cc9c87ce4270649c207d

**Comment by JoelSpeed** [comment]
@yyoshiki41 Could you please rebase this and then I'll give it another review

**Comment by yyoshiki41** [comment]
@JoelSpeed 
rebased and move options to follow bellow pr
https://github.com/oauth2-proxy/oauth2-proxy/pull/489

Thanks!

File: providers/github_test.go [hunk 8]
```diff
@@ -318,3 +318,78 @@ func TestGitHubProviderGetUserNameWithRepoAndTokenWithoutPushAccess(t *testing.T
...
+func TestGitHubProviderGetEmailAddressWithUsername(t *testing.T) {
+	b := testGitHubBackend(map[string][]string{
+		"/user":        {`{"email": "michael.bland@gsa.gov", "login": "mbland"}`},
+		"/user/emails": {`[ {"email": "michael.bland@gsa.gov", "verified": true, "primary": true} ]`},
+	})
+	defer b.Close()
+
+	bURL, _ := url.Parse(b.URL)
+	p := testGitHubProvider(bURL.Host)
+	p.SetUsers([]string{"mbland", "octcat"})
```
Comment (JoelSpeed) [inline providers/github_test.go:331]:
```suggestion
	p.SetUsers([]string{"mbland", "octocat"})
```

File: providers/github_test.go [hunk 9]
```diff
@@ -318,3 +318,78 @@ func TestGitHubProviderGetUserNameWithRepoAndTokenWithoutPushAccess(t *testing.T
...
+func TestGitHubProviderGetEmailAddressWithNotAllowedUsername(t *testing.T) {
+	b := testGitHubBackend(map[string][]string{
+		"/user":        {`{"email": "michael.bland@gsa.gov", "login": "mbland"}`},
+		"/user/emails": {`[ {"email": "michael.bland@gsa.gov", "verified": true, "primary": true} ]`},
+	})
+	defer b.Close()
+
+	bURL, _ := url.Parse(b.URL)
+	p := testGitHubProvider(bURL.Host)
+	p.SetUsers([]string{"octcat"})
```
Comment (JoelSpeed) [inline providers/github_test.go:348]:
```suggestion
	p.SetUsers([]string{"octocat"})
```

**Comment by yyoshiki41** [comment]
@JoelSpeed 

fixed to improve readability in the docs
https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/0333f9f414c96fb9c34725fd10d506fcf4416e65

sorry, I missed your review.
I fixed it by https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/86cadbfbc6b4edcdb1981e0a561188c86caf3f5b

I fixed all your review points .
Thanks.

File: providers/github.go [hunk 10]
```diff
@@ -80,6 +82,11 @@ func (p *GitHubProvider) SetRepo(repo, token string) {
 	p.Token = token
 }
 
+// SetUsers configures allowed usernames
+func (p *GitHubProvider) SetUsers(users []string) {
+	p.Users = append(p.Users, users...)
```
[the remark this answers is not in the document]
Comment (yyoshiki41) [reply providers/github.go]: Sorry, I missed this points.
fixed by 86cadbf

File: pkg/apis/options/options.go [hunk 11]
```diff
@@ -228,6 +229,7 @@ func NewFlagSet() *pflag.FlagSet {
 	flagSet.String("github-team", "", "restrict logins to members of this team")
 	flagSet.String("github-repo", "", "restrict logins to collaborators of this repository")
 	flagSet.String("github-token", "", "the token to use when verifying repository collaborators (must have push access to the repository)")
+	flagSet.StringSlice("github-user", []string{}, "allowed users with these usernames to login even if they do not belong to the specified org and team or collaborators (may be given multiple times)")
```
Comment (JoelSpeed) [inline pkg/apis/options/options.go:232]:
```suggestion
	flagSet.StringSlice("github-user", []string{}, "allow users with these usernames to login even if they do not belong to the specified org and team or collaborators (may be given multiple times)")
```

**Pull request merged**